File: src/scanner.py
from typing import List, Any

from token import (
    Token,
    TokenType,
    SINGLE_CHARACTER_TOKENS,
    ONE_OR_MORE_CHARACTER_TOKENS,
    WHITESPACE,
    STRING_STARTERS,
    KEYWORDS,
    COMMENT_TOKENS
)


class Scanner:
    def __init__(self, source: str) -> None:
        self.source = source
        self.tokens: List[Token] = []
        self.start = 0
        self.current = 0
        self.line = 1
        # Source offset where the current line began -- lets column_at() turn
        # an absolute offset into a 1-indexed column within that line.
        self.line_start = 0
        # Line/column of the token currently being scanned, captured once at
        # the top of scan_tokens()'s loop -- used (rather than self.line/
        # column_at(self.start) at add_token() time) so a token that itself
        # spans multiple lines (e.g. a multi-line string) is still reported
        # at the position it *started*, not wherever scanning ended up.
        self.start_line = 1
        self.start_column = 1

    @property
    def current_token(self) -> str:
        return self.source[self.start:self.current]

    def column_at(self, offset: int) -> int:
        return offset - self.line_start + 1

    def advance_newline(self) -> None:
        # Call right after advance() has consumed a '\n', so self.current
        # already points just past it -- keeps self.line_start/self.line in
        # sync wherever a newline is consumed outside the main scan_token()
        # dispatch (inside strings and block comments).
        self.line += 1
        self.line_start = self.current

    def is_at_end(self) -> bool:
        return self.current >= len(self.source)

    def advance(self) -> str:
        self.current += 1
        return self.current_token

    def add_token(self, token_type: TokenType, literal: Any = None) -> None:
        self.tokens.append(Token(token_type, self.current_token, literal, self.start_line, self.start_column))

    def match(self, expected: str) -> bool:
        if self.is_at_end() or self.source[self.current] != expected:
            return False
        self.current += 1
        return True

    def peek(self) -> str:
        if self.is_at_end():
            return '\0'
        return self.source[self.current]

    def peek_next(self) -> str:
        if self.current + 1 >= len(self.source):
            return '\0'
        return self.source[self.current + 1]

    def is_digit(self, c: str = None) -> bool:
        if c is None:
            c = self.peek()
        return '0' <= c <= '9'

    def is_alpha(self, c: str = None) -> bool:
        if c is None:
            c = self.peek()
        return ('a' <= c <= 'z') or ('A' <= c <= 'Z') or c == '_'

    def is_alpha_numeric(self, c: str = None) -> bool:
        return self.is_alpha(c) or self.is_digit(c)

    def is_recognized(self, c: str) -> bool:
        # Anything scan_token()'s dispatch would actually know what to do
        # with -- used to find where a run of otherwise-unrecognized
        # characters ends (see the fallback branch in scan_token()).
        return (
            c in SINGLE_CHARACTER_TOKENS
            or c in ONE_OR_MORE_CHARACTER_TOKENS
            or c in WHITESPACE
            or c in STRING_STARTERS
            or c == TokenType.NEWLINE
            or self.is_digit(c)
            or self.is_alpha(c)
        )

    def string(self, char: str) -> None:
        while self.peek() != char and not self.is_at_end():
            if self.peek() == '\n':
                self.advance()
                self.advance_newline()
            else:
                self.advance()
        if self.is_at_end():
            import iqalox
            # Reports at the opening quote's own position (not wherever
            # end-of-file was actually reached), since that's where the user
            # needs to look to fix it.
            iqalox.Iqalox.error(
                Token(TokenType.STRING, char, None, self.start_line, self.start_column), "Unterminated string."
            )
            return
        self.advance()
        value = self.source[self.start + 1:self.current - 1]
        self.add_token(TokenType.STRING, value)

    def number(self) -> None:
        while self.is_digit():
            self.advance()
        if self.peek() == '.' and self.is_digit():
            self.advance()
            while self.is_digit():
                self.advance()
        self.add_token(TokenType.NUMBER, float(self.source[self.start:self.current]))

    def identifier(self) -> None:
        while self.is_alpha_numeric():
            self.advance()
        text = self.current_token
        token_type = KEYWORDS.get(text, TokenType.IDENTIFIER)
        self.add_token(token_type)

    def scan_token(self) -> None:
        c = self.advance()
        if c in SINGLE_CHARACTER_TOKENS:
            self.add_token(TokenType(c))
        elif c in ONE_OR_MORE_CHARACTER_TOKENS:
            compounds = [i for i in ONE_OR_MORE_CHARACTER_TOKENS if i.startswith(c) and len(i) > 1]
            token = c
            for compound in compounds:
                if self.match(compound[1]):
                    token = compound
                    break
            if token in COMMENT_TOKENS:
                if token == TokenType.COMMENT:
                    while self.peek() != '\n' and not self.is_at_end():
                        self.advance()
                elif token == TokenType.BLOCK_COMMENT_START:
                    while not (self.peek() == '#' and self.peek_next() == '>') and not self.is_at_end():
                        if self.peek() == '\n':
                            self.advance()
                            self.advance_newline()
                        else:
                            self.advance()
                    self.advance()
                    self.advance()
            else:
                try:
                    self.add_token(TokenType(token))
                except ValueError:
                    # `token` didn't extend into a longer valid token (e.g. a
                    # bare '|' not followed by '>') and isn't valid on its own.
                    import iqalox
                    iqalox.Iqalox.error(
                        Token(TokenType.NULL_CHAR, c, None, self.start_line, self.start_column),
                        f"Unexpected character: {c}",
                    )
        elif c in WHITESPACE:
            return
        elif c == TokenType.NEWLINE:
            self.add_token(TokenType.SEMICOLON)
            self.advance_newline()
        elif c in STRING_STARTERS:
            self.string(c)
        elif self.is_digit(c):
            self.number()
        elif self.is_alpha(c):
            self.identifier()
        else:
            # A character with no recognized meaning on its own. Keep
            # consuming as long as what follows is *also* unrecognized, so a
            # whole run of garbage characters (e.g. `@@@`) is reported as one
            # error instead of one per character.
            while not self.is_at_end() and not self.is_recognized(self.peek()):
                self.advance()
            import iqalox
            iqalox.Iqalox.error(
                Token(TokenType.NULL_CHAR, self.current_token, None, self.start_line, self.start_column),
                f"Unexpected character{'s' if len(self.current_token) > 1 else ''}: {self.current_token}",
            )

    def scan_tokens(self) -> List[Token]:
        while not self.is_at_end():
            self.start = self.current
            self.start_line = self.line
            self.start_column = self.column_at(self.start)
            self.scan_token()
        self.tokens.append(Token(TokenType.EOF, "", None, self.line, self.column_at(self.current)))
        return self.tokens

File: compiler/src/Scanner.fs
/// Scanner: source text -> `Token` list.
///
/// Ports `poc/src/scanner.py`'s design, including its two proven-correct
/// post-release fixes (accurate line/column tracking via a `lineStart`
/// offset; coalescing a run of unrecognized characters into a single
/// error instead of one per character), plus escape sequences in string
/// literals (new for `0.1`, see `docs/PLAN-0.1.md` decision 5).
///
/// This is a fresh implementation, not a line-for-line port, because
/// writing it carefully surfaced several `poc` bugs that were never
/// "proven correct" and so aren't carried forward:
///   - Decimal number literals never worked in `poc` at all (`number()`
///     compares `self.peek()` against a digit check that also defaults to
///     `self.peek()`, i.e. it always compares the same '.' character to
///     itself and never looks at what follows it) -- `3.14` scans as
///     `NUMBER("3")`, `DOT`, `NUMBER("14")`. Fixed here by checking the
///     character *after* the dot.
///   - A leading-underscore identifier (`_foo`) misscans as a bare `_`
///     (the ignore operator) followed by a separate `foo` identifier,
///     since `_` is checked as its own token before the identifier
///     dispatch runs. Fixed here by checking whether an alphanumeric
///     character follows before deciding `_` is standalone.
///   - The `...` (ellipsis) token under-consumes by one character (`poc`'s
///     generic compound-matching only ever advances one extra character
///     regardless of the compound's actual length) -- moot in `poc` since
///     `...` has no grammar yet, but worth not carrying forward. Fixed
///     here by a longest-match search that consumes exactly the matched
///     lexeme's length.
///   - A bare `#>` outside a block comment is silently swallowed with no
///     token or error (an accident of treating comment markers as just
///     another entry in the general operator-compound table). Fixed here
///     by giving comments their own explicit dispatch instead.
///   - An unterminated block comment (`<#` with no matching `#>` before
///     EOF) produces no error at all in `poc`. Fixed here to report one,
///     matching how an unterminated string is already handled.
module Iqalox.Scanner

open System.Text
open Iqalox.Token

type ScanError = { Message: string; Line: int; Column: int }

/// Multi- and single-character punctuation, longest lexeme first so a
/// longest-match search finds e.g. `...` before `.`, or `??` before `?`.
/// Comment markers (`#`, `<#`, `#>`) and string quotes are deliberately
/// not here -- `Scanner`'s main dispatch handles those explicitly instead
/// of through this generic table (see the module doc comment).
let private operatorCandidates: (string * TokenType) list =
    [ "...", Ellipsis
      "!=", BangEqual
      "==", EqualEqual
      ">=", GreaterEqual
      "<=", LessEqual
      "--", MinusMinus
      "->", Arrow
      "<-", LeftArrow
      "++", PlusPlus
      "?:", QuestionMarkColon
      "??", DoubleQuestionMark
      "|>", Pipe
      "(", LeftParen
      ")", RightParen
      "{", LeftBrace
      "}", RightBrace
      "[", LeftBracket
      "]", RightBracket
      ",", Comma
      ";", Semicolon
      "/", Slash
      "\\", Backslash
      "*", Star
      "%", Percent
      "^", Power
      "!", Bang
      "=", Equal
      ">", Greater
      "<", Less
      "-", Minus
      "+", Plus
      ".", Dot
      "?", QuestionMark
      ":", Colon
      "|", VerticalBar ]
    |> List.sortByDescending (fun (lexeme, _) -> lexeme.Length)

let private operatorStartChars: Set<char> =
    operatorCandidates |> List.map (fun (lexeme, _) -> lexeme.[0]) |> Set.ofList

let private isDigit (c: char) = c >= '0' && c <= '9'
let private isAlpha (c: char) = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c = '_'
let private isAlphaNumeric (c: char) = isAlpha c || isDigit c

/// Recognized escape sequences (decision 5, docs/PLAN-0.1.md): the common
/// C-family set. Anything else after a backslash is a hard error -- no
/// silent fallback to a literal backslash-plus-character.
let private escapeSequences: Map<char, char> =
    Map.ofList [ 'n', '\n'; 't', '\t'; 'r', '\r'; '\\', '\\'; '\'', '\''; '"', '"'; '0', '\000' ]

type private ScannerState(source: string) =
    let tokens = ResizeArray<Token>()
    let errors = ResizeArray<ScanError>()
    let mutable start = 0
    let mutable current = 0
    let mutable line = 1
    let mutable lineStart = 0
    // Line/column of the token currently being scanned, captured once per
    // token (mirrors `poc`'s `start_line`/`start_column`) so a token that
    // itself spans multiple lines (a multi-line string, a block comment)
    // is still reported at the position it *started*, not wherever
    // scanning ended up.
    let mutable startLine = 1
    let mutable startColumn = 1

    let isAtEnd () = current >= source.Length

    let peek () = if isAtEnd () then '\000' else source.[current]

    let peekNext () =
        if current + 1 >= source.Length then '\000' else source.[current + 1]

    let advance () =
        let c = source.[current]
        current <- current + 1
        c

    let columnAt (offset: int) = offset - lineStart + 1

    /// Call right after `advance()` has consumed a '\n', so `current`
    /// already points just past it.
    let advanceNewline () =
        line <- line + 1
        lineStart <- current

    let currentLexeme () = source.Substring(start, current - start)

    let addToken (tokenType: TokenType) (literal: Literal) =
        tokens.Add(
            { Type = tokenType
              Lexeme = currentLexeme ()
              Literal = literal
              Line = startLine
              Column = startColumn }
        )

    let addError (message: string) (errLine: int) (errColumn: int) =
        errors.Add { Message = message; Line = errLine; Column = errColumn }

    let isRecognized (c: char) =
        c = '\n'
        || c = ' '
        || c = '\t'
        || c = '\r'
        || c = '#'
        || c = '\''
        || c = '"'
        || isDigit c
        || isAlpha c
        || Set.contains c operatorStartChars

    let scanUnrecognizedRun () =
        while not (isAtEnd ()) && not (isRecognized (peek ())) do
            advance () |> ignore
        let lexeme = currentLexeme ()
        let suffix = if lexeme.Length > 1 then "s" else ""
        addError $"Unexpected character{suffix}: {lexeme}" startLine startColumn

    let tryMatchOperator () : TokenType option =
        let remaining = source.Length - start
        operatorCandidates
        |> List.tryFind (fun (lexeme, _) -> lexeme.Length <= remaining && source.Substring(start, lexeme.Length) = lexeme)
        |> Option.map (fun (lexeme, tokenType) ->
            current <- start + lexeme.Length
            tokenType)

    let scanLineComment () =
        while peek () <> '\n' && not (isAtEnd ()) do
            advance () |> ignore

    let scanBlockComment () =
        // `current` is already past "<#" by the time this is called.
        let mutable closed = false
        while not closed && not (isAtEnd ()) do
            if peek () = '#' && peekNext () = '>' then
                advance () |> ignore
                advance () |> ignore
                closed <- true
            elif peek () = '\n' then
                advance () |> ignore
                advanceNewline ()
            else
                advance () |> ignore
        if not closed then
            addError "Unterminated block comment." startLine startColumn

    let scanString (quote: char) =
        let value = StringBuilder()
        let mutable finished = false
        // Distinct from `finished`: on a bad escape we keep scanning to
        // find the real closing quote (so scanning resumes at the right
        // place afterward) but stop producing a String token, and only
        // report the first bad escape in a given string rather than one
        // per occurrence -- same "don't spam" philosophy as coalescing a
        // run of unrecognized characters elsewhere in this scanner.
        let mutable hadError = false
        while not finished do
            if isAtEnd () then
                addError "Unterminated string." startLine startColumn
                finished <- true
            elif peek () = quote then
                advance () |> ignore
                finished <- true
            elif peek () = '\\' then
                let backslashColumn = columnAt (current)
                advance () |> ignore // consume the backslash
                if isAtEnd () then
                    addError "Unterminated string." startLine startColumn
                    finished <- true
                else
                    let escapeChar = advance ()
                    match Map.tryFind escapeChar escapeSequences with
                    | Some resolved -> value.Append(resolved) |> ignore
                    | None ->
                        if not hadError then
                            addError $"Unrecognized escape sequence: \\{escapeChar}" line backslashColumn
                        hadError <- true
            elif peek () = '\n' then
                value.Append('\n') |> ignore
                advance () |> ignore
                advanceNewline ()
            else
                value.Append(advance ()) |> ignore
        if not hadError then
            addToken String (StringLiteral(value.ToString()))

    let scanNumber () =
        while isDigit (peek ()) do
            advance () |> ignore
        if peek () = '.' && isDigit (peekNext ()) then
            advance () |> ignore // consume the '.'
            while isDigit (peek ()) do
                advance () |> ignore
        addToken Number (NumberLiteral(float (currentLexeme ())))

    let scanIdentifierOrKeyword () =
        while isAlphaNumeric (peek ()) do
            advance () |> ignore
        let text = currentLexeme ()
        match Map.tryFind text keywords with
        | Some tokenType -> addToken tokenType NoLiteral
        | None -> addToken Identifier NoLiteral

    let scanToken () =
        let c = advance ()
        if c = '\n' then
            addToken Semicolon NoLiteral
            advanceNewline ()
        elif c = ' ' || c = '\t' || c = '\r' then
            ()
        elif c = '#' then
            scanLineComment ()
        elif c = '<' && peek () = '#' then
            advance () |> ignore
            scanBlockComment ()
        elif c = '\'' || c = '"' then
            scanString c
        elif isDigit c then
            scanNumber ()
        elif c = '_' then
            if isAlphaNumeric (peek ()) then
                scanIdentifierOrKeyword ()
            else
                addToken Underscore NoLiteral
        elif isAlpha c then
            scanIdentifierOrKeyword ()
        else
            match tryMatchOperator () with
            | Some tokenType -> addToken tokenType NoLiteral
            | None -> scanUnrecognizedRun ()

    member _.ScanTokens() : Token list * ScanError list =
        while not (isAtEnd ()) do
            start <- current
            startLine <- line
            startColumn <- columnAt start
            scanToken ()

        tokens.Add(
            { Type = Eof
              Lexeme = ""
              Literal = NoLiteral
              Line = line
              Column = columnAt current }
        )

        List.ofSeq tokens, List.ofSeq errors

/// Scans `source` into a list of tokens plus any scan errors. Scanning
/// never throws -- a bad character or an unterminated string/comment is
/// recorded as a `ScanError` and scanning continues, matching `poc`'s
/// "report and keep going" approach (mirroring parse-error recovery, not
/// an exception-per-error design).
let scanTokens (source: string) : Token list * ScanError list =
    ScannerState(source).ScanTokens()
